src/asl.ts

```diff
@@ -5526,28 +5253,1768 @@ export namespace ASLGraph {
   }
 }
 
-export namespace ASL {
-  export function isTruthy(v: string): Condition {
-    return and(
-      isPresentAndNotNull(v),
-      or(
-        and(isString(v), not(stringEquals(v, ""))),
-        and(isNumeric(v), not(numericEquals(v, 0))),
-        and(isBoolean(v), booleanEquals(v, true))
-      )
+/**
+ * Operations that can help "optimize" an ASL state machine.
+ */
+namespace ASLOptimizer {
+  /**
+   * Traverses the graph and removes any nodes unreachable from the start state.
+   */
+  export function removeUnreachableStates(
+    startState: string,
+    states: States
+  ): States {
+    const reachableStates = findReachableStates(startState, states);
+
+    // only take the reachable states
+    return Object.fromEntries(
+      Object.entries(states).filter(([name]) => reachableStates.has(name))
     );
   }
 
-  export function and(...cond: (Condition | undefined)[]): Condition {
-    const conds = cond.filter((c): c is Condition => !!c);
-    return conds.length > 1
-      ? {
-          And: conds,
-        }
-      : conds.length === 0
-      ? ASL.trueCondition()
-      : conds[0]!;
-  }
+  /**
+   * Given a directed adjacency matrix, return a `Set` of all reachable states from the start state.
+   */
+  function findReachableStates(
+    startState: string,
+    states: Record<string, State>
+  ) {
+    const visited = new Set<string>();
+
+    // starting from the start state, find all reachable states
+    depthFirst(startState);
+
+    return visited;
+
+    function depthFirst(stateName: string) {
+      if (visited.has(stateName)) return;
+      visited.add(stateName);
+      const state = states[stateName]!;
+      visitTransition(state, depthFirst);
+    }
+  }
+
+  /**
+   * Find all {@link Pass} states that do not do anything.
+   */
+  function findEmptyStates(startState: string, states: States): string[] {
+    return Object.entries(states)
+      .filter((entry): entry is [string, Pass] => {
+        const [name, state] = entry;
+        return (
+          name !== startState &&
+          isPassState(state) &&
+          !!state.Next &&
+          !(
+            state.End ||
+            state.InputPath ||
+            state.OutputPath ||
+            state.Parameters ||
+            state.Result ||
```

Comment:
What if this was an empty string or 0 -ie a falsy value

Reply:
(this is existing code I moved)

I don't think any would matter as falsey. For input path, result, result selector, or parameters to matter, ResultPath would need to be truthy. Output path cannot be null.

```
Result: false,
ResultPath: undefined
```
^ does nothing.

Reply:
Why does ResultPath matter? If Result: false and ResultPath: $.a what would happen

Reply:
```ts
            state.End ||
            state.InputPath ||
            state.OutputPath ||
            state.Parameters ||
            state.Result ||
            state.ResultPath
```

`true`?